Copyright Jana Schaich Borg/Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)

# MySQL 연습문제 9: Subquery와 파생 테이블

SubQuery: 쿼리 안에 있는 작은 쿼리
Subquery는 SELECT, WHERE과 FROM절에서 사용되며, Subquery가 FROM절에 포함이 되면 파생 테이블이라고 부름  

### Subquery를 사용하는 이유: 

+ 원하는 정보를 출력하는 가장 논리적인 방법
+ 길고 복잡한 쿼리의 문제를 고치는데 도움이 됨
+ JOIN하는 것보다 빠르게 실행됨

### Subquery는 괄호안에 들어가 있어야 합니다. Join과는 다른 Subquery의 규칙은 다음과 같습니다: 

+ ORDER BY절은 Subquery안에서 사용할 수 없음
+ SELECT 또는 WHERE절에 있는 Subquery는 한 열 이상을 출력 할 수 없음. 두개 이상의 열을 출력하는 SELECT 또는 WHERE 절의 Subquery는 IN 연산자와 같이 사용해야 함.

In [5]:
%load_ext sql
%sql mysql://studentuser:studentpw@localhost/dognitiondb
%sql USE dognitiondb

The sql extension is already loaded. To reload it, use:
  %reload_ext sql
 * mysql://studentuser:***@localhost/dognitiondb
0 rows affected.


[]

## 1) Subquery는 계산 할때 유용합니다

평균보다 긴 기간 (duration)에 관련된 데이터만 출력하고 싶으면 Subquery를 사용해서 평균 행을 구한 다음, 그 조건을 Where절에 포함시키고 구해진 평균 값보다 긴 행만 출력하라고 하면 됨. 

```mysql
SELECT *
FROM exam_answers 
WHERE TIMESTAMPDIFF(minute,start_time,end_time) >
    (SELECT AVG(TIMESTAMPDIFF(minute,start_time,end_time)) AS AvgDuration
     FROM exam_answers
     WHERE TIMESTAMPDIFF(minute,start_time,end_time)>0);
```

<주의해야 할 점>
+ Subquery안에 있는 TIMESTAMPDIFF는 Subquery밖에 있는 TIMESTAMP와 비교됨.
+ Subquery를 쉽게 읽으려면 Subquery부분을 키보드의 Tab키를 사용해서 Indent하면 됨.

**질문 1: Subquery를 사용해서 테스트 기간이 (test duration) Yawn Warm-up 게임의 평균 테스트 기간보다 긴 데이터를 exam_answers테이블 에서 출력하려고 합니다. 그렇게 하기 위해서 먼저 Yawn Warm-up 게임의 평균 기간을 출력하는 쿼리를 쓰세요. 데이터에 문제가 있기 때문에 기간이 음수 (negative value)인 행들을 제외 해야합니다. (쿼리를 올바르게 실행하면 평균값으로 9934이 나와야 합니다) **

In [6]:
%sql SELECT AVG(TIMESTAMPDIFF(minute,start_time,end_time)) FROM exam_answers\
WHERE TIMESTAMPDIFF(minute, start_time, end_time) > 0 AND test_name="Yawn Warm-up";

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(Decimal('9933.5197'),)]

**질문 2: Subquery가 될 쿼리를 잘 쓴것을 확인했으니 이 subquery를 메인 쿼리 안에 넣어서 테스트 기간이 (test duration) Yawn Warm-up 게임의 평균 테스트 기간보다 오래 걸리는 데이터를 exam_answers테이블 에서 출력하세요. (11059행이 출력되야 합니다):**

In [14]:
%sql SELECT count(*) FROM exam_answers WHERE TIMESTAMPDIFF(minute, start_time, end_time) > \
    (SELECT AVG(TIMESTAMPDIFF(minute, start_time, end_time)) FROM exam_answers WHERE TIMESTAMPDIFF(minute, start_time, end_time) > 0 AND test_name="Yawn Warm-up");

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(11059,)]

쿼리를 잘 썼나 확인하려면 Subquery가 들어가는 자리에 숫자 "9934"를 넣고 같은 값이 나오는지 보면 됨. 이렇게 항상 체크하는 습관을 들이는 것이 중요.

위의 쿼리를 통해서 subquery를 사용하면 특정 숫자나 이름을 적지 않고 Where에 있는 조건을 유연하게 지정할 수 있음. 이 기능은 실시간 정보를 출력하는 대시보드를 만들어야 할때 도움이 됨.
   
## 2) Subquery는 멤버쉽을 테스트할때 유용합니다 

Subquery는 출력된 데이터의 행이 다른 그룹의 행에 속하는지 알아내는데 도움 됨. 이렇게 사용하려면 IN, NOT IN, EXISTS, and NOT EXISTS 연산자를 사용해야 함. 
하단 세 코드는 모두 동일함
```mysql
SELECT * 
FROM users
WHERE state IN ('NC','NY');
```
```mysql
SELECT * 
FROM users
WHERE state ='NC' OR state ='NY';
```
```mysql
SELECT * 
FROM users
WHERE state NOT IN ('NC','NY');
```
**질문 3: IN 연산자를 사용하여서 exam_answers테이블에 하위 카테고리(subcategory)가 "Puzzles", "Numerosity"나 "Bark Game"의 게임에 해당하는 행이 몇 개가 있는지 알아보세요. 총 163022개의 행이 출력되야 합니다**

In [18]:
%sql SELECT COUNT(*) FROM exam_answers WHERE subcategory_name IN ("Puzzles","Numerosity","Bark Game");

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(163022,)]

**질문 4: NOT IN 연산자를 사용해서 dog테이블에 몇 마리의 고유의 개가 있는지 구하는데, 품종 그룹 (breed_group)이 "Working", "Sporting"나 "Herding"가 아닌 행만 출력하세요. 답은 7961행입니다.**

In [24]:
%sql SELECT COUNT(*) FROM dogs WHERE breed_group NOT IN ("Working","Sporting","Herding");

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(7961,)]

EXISTS와 NOT EXISTS연산자는 IN과 NOT IN연산자와 비슷한 기능을 가지고 있지만 Subquery안에서만 사용될 수 있음. EXISTS와 NOT EXISTS는 논리적인 문장이라서 원시 자료를 출력하는 것이 아니라 논리에 따라 TRUE나 FALSE로 출력. 

EXISTS와 Subquery를 같이 쓰면 users와 dogs테이블에 있는 사용자에 대한 정보를 비교할 수 있습니다. dogs 테이블과 users테이블에 있는 사용자들의 목록을 출력하려면 다음과 같은 쿼리를 작성하면 됩니다. 

```sql
SELECT DISTINCT u.user_guid AS uUserID
FROM users u
WHERE EXISTS (SELECT d.user_guid
              FROM dogs d 
              WHERE u.user_guid =d.user_guid);
```

Subquery안에서 user_guid가 아닌 "*" (별표) 를 써도 같은 결과를 출력할 수 있음:
```sql
SELECT DISTINCT u.user_guid AS uUserID
FROM users u
WHERE EXISTS (SELECT *
              FROM dogs d 
              WHERE u.user_guid =d.user_guid);
```

**질문 5: NOT EXISTS절을 사용하여 dogs 테이블에 없는 users 테이블의 고유 사용자 수를 구하세요. MySQL 연습 문제 8의 질문 10과 같은 답인 2226행을 출력해야 합니다. **

#### ** subquery 안에서는 바깥에서 쓴 alias 사용 가능한건지?

In [12]:
%sql SELECT COUNT(u.user_guid) FROM users u WHERE NOT EXISTS (\
    SELECT d.user_guid FROM dogs d WHERE u.user_guid=d.user_guid);

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(2226,)]

## 3) Subquery로 정확한 논리로 테이블을 만들 수 있고 파생된 테이블을 출력할 수 있습니다 

Subquery가 유용한 세번째 이유는 단순한 JOIN보다 더 정확한 논리의 결과를 출력할 수 있기 때문임. 

중복 된 행을 여러 개 결합하고 나중에 GROUP BY 절로 결과를 그룹화하는 대신, 먼저 중복된 행을 subqueyr로 고유값만 추출후 JOIN하여 간단하게 처리 가능함.

```mysql 
SELECT DistinctUUsersID.user_guid AS uUserID, d.user_guid AS dUserID, count(*) AS numrows
FROM (SELECT DISTINCT u.user_guid 
      FROM users u) AS DistinctUUsersID 
LEFT JOIN dogs d
   ON DistinctUUsersID.user_guid=d.user_guid
GROUP BY DistinctUUsersID.user_guid
ORDER BY numrows DESC
``` 

**한번 직접 해보세요:**

In [9]:
%sql SELECT DISTINCTUUsersID.user_guid AS uUserID, d.user_guid AS duserID, count(*) AS numrows\
FROM (SELECT DISTINCT u.user_guid FROM users u) AS DISTINCTUUsersID LEFT JOIN dogs d on DISTINCTUUsersID.user_guid =d.user_guid\
GROUP BY DISTINCTUUsersID.user_guid ORDER BY numrows DESC LIMIT 10;

 * mysql://studentuser:***@localhost/dognitiondb
10 rows affected.


[('ce7b75bc-7144-11e5-ba71-058fbc01cf0b', 'ce7b75bc-7144-11e5-ba71-058fbc01cf0b', 1819),
 ('ce225842-7144-11e5-ba71-058fbc01cf0b', 'ce225842-7144-11e5-ba71-058fbc01cf0b', 26),
 ('ce2258a6-7144-11e5-ba71-058fbc01cf0b', 'ce2258a6-7144-11e5-ba71-058fbc01cf0b', 20),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'ce135e14-7144-11e5-ba71-058fbc01cf0b', 13),
 ('ce29675e-7144-11e5-ba71-058fbc01cf0b', 'ce29675e-7144-11e5-ba71-058fbc01cf0b', 11),
 ('ce134492-7144-11e5-ba71-058fbc01cf0b', 'ce134492-7144-11e5-ba71-058fbc01cf0b', 9),
 ('ce6676d0-7144-11e5-ba71-058fbc01cf0b', 'ce6676d0-7144-11e5-ba71-058fbc01cf0b', 8),
 ('ce83d2ca-7144-11e5-ba71-058fbc01cf0b', 'ce83d2ca-7144-11e5-ba71-058fbc01cf0b', 8),
 ('ce32305a-7144-11e5-ba71-058fbc01cf0b', 'ce32305a-7144-11e5-ba71-058fbc01cf0b', 7),
 ('ce7adeea-7144-11e5-ba71-058fbc01cf0b', 'ce7adeea-7144-11e5-ba71-058fbc01cf0b', 7)]

**<mark> Subquery가 있는 쿼리들은 항상 가장 안에 있는 Subquery를 먼저 실행하고 바깥에 있는 쿼리를 실행합니다. </mark>**
  
**이 쿼리 구문에 대해서 알아야 할 중요한 점이 있습니다.** 
+ 첫번째, subquery가 파생시킨 테이블에 "DistinctUUsersID"라는 임시 이름을 지정했는데, FROM절 뒤에 오는 subquery에는 항상 임시 이름을 지정해줘야 함. 
   
+ 두번째, *파생된 테이블의 정보를 사용하려면 항상 이 임시 이름을 사용해야 함*  

+ 세번째, subquery안에서는 subquery밖의 테이블의 임시 이름을 사용 할 수 있지만, subquery밖에 있는 쿼리에서는 쿼리 결과에 출력되지 않은 subquery안의 임시 이름을 사용할 수 없음.

```sql
SELECT u.user_guid AS uUserID, d.user_guid AS dUserID, count(*) AS numrows
FROM (SELECT DISTINCT u.user_guid 
      FROM users u) AS DistinctUUsersID 
LEFT JOIN dogs d
   ON DistinctUUsersID.user_guid=d.user_guid
GROUP BY DistinctUUsersID.user_guid
ORDER BY numrows DESC
```

"u"는 subquery안에서 user테이블에게 지정된 임시 이름이기 때문에 위의 쿼리는 실행되지 않습니다.  **위의 쿼리를 실행해보고 어떤 오류 메시지가 뜨는지 살펴보세요:**

In [13]:
%sql SELECT u.user_guid AS uUserID, d.user_guid AS dUserID, count(*) AS numrows FROM (SELECT DISTINCT u.user_guid FROM users u)\
AS DistinctUUsersID LEFT JOIN dogs d\
ON DistinctUUsersID.user_guid=d.user_guid\
GROUP BY DistinctUUsersID.user_guid\
ORDER BY numrows DESC LIMIT 10;

 * mysql://studentuser:***@localhost/dognitiondb
(MySQLdb._exceptions.OperationalError) (1054, "Unknown column 'u.user_guid' in 'field list'")
[SQL: SELECT u.user_guid AS uUserID, d.user_guid AS dUserID, count(*) AS numrows FROM (SELECT DISTINCT u.user_guid FROM users u) AS DistinctUUsersID LEFT JOIN dogs d ON DistinctUUsersID.user_guid=d.user_guid GROUP BY DistinctUUsersID.user_guid ORDER BY numrows DESC LIMIT 10;]
(Background on this error at: http://sqlalche.me/e/e3q8)


#### -> Unknown column 'u.user_guid' in 'field list 에러 발생

FROM절 뒤에 Subquery를 쓰면 쿼리를 실행하는것이 오래 걸릴 수도 있음.파생절 사용은 유의해서 사용해야함.
   
# 이제 이론을 숙지했으니 Subquery들을 더 연습해봅시다!

**질문 6: Working, Sporting, Herding 품종 그룹 (breed group)에 속하는 고유의 개의 dog ID (dog_guid), 품종 그룹 (breed group), 개 주인이 살고 있는 주 이름 (state of the owner), 개 주인의 우편번호 (zip of the owner)를 출력하는 쿼리를 IN과 equijoin 구문을 써서 출력하시오. (답은 10,254행입니다. 이 쿼리가 실행되는 것에는 다른 쿼리보다 오래 걸릴 것입니다.)**

In [20]:
%sql SELECT COUNT(*) FROM (SELECT DISTINCT d.breed_group, d.dog_guid, u.state, u.zip FROM dogs d,users u WHERE u.user_guid=d.user_guid AND d.breed_group IN ('Working','Sporting','Herding')) c;

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(10254,)]

**질문 7: 위와 같은 쿼리를 JOIN 구문을 사용해서 써보세요.**

In [21]:
%sql SELECT COUNT(*) FROM (SELECT DISTINCT d.breed_group, d.dog_guid, u.state, u.zip FROM dogs d JOIN users u ON u.user_guid=d.user_guid WHERE d.breed_group IN ('Working','Sporting','Herding')) c;

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(10254,)]

**질문 8: NOT EXISTS를 사용해서 users테이블에 없는 dogs테이블의 사용자들의 목록을 출력하세요. user_guid와 dog_guid 열들을 출력하세요 (답으로 2개의 행이 출력되어야 합니다)**

In [24]:
%sql SELECT d.user_guid, d.dog_guid FROM dogs d\
WHERE NOT EXISTS (SELECT u.user_guid FROM users u WHERE u.user_guid=d.user_guid);

 * mysql://studentuser:***@localhost/dognitiondb
2 rows affected.


[(None, 'fd7c0a66-7144-11e5-ba71-058fbc01cf0b'),
 (None, 'fdbb6b7a-7144-11e5-ba71-058fbc01cf0b')]

**질문 9: 전에 실행했던 쿼리를 생각해보면 dogs테이블과 left outer join을 했는데도 'ce7b75bc-7144-11e5-ba71-058fbc01cf0b'라는 user_guid를 가진 사용자에 대한 행이 1819개 출력됬었습니다. 왜 그랬을까 자세히 살펴보면 dogs테이블에도 중복되는 user_guids가 있기 때문입니다. 전에 썼던 쿼리를 수정해서 users테이블에서 고유한 user ID만 출력해서 dogs테이블에 Join하려면 어떻게 해야 할까요?**

다음은 전에 썼던 쿼리입니다:

```sql
SELECT DistinctUUsersID.user_guid AS uUserID, d.user_guid AS dUserID, count(*) AS numrows
FROM (SELECT DISTINCT u.user_guid 
      FROM users u) AS DistinctUUsersID 
LEFT JOIN dogs d
  ON DistinctUUsersID.user_guid=d.user_guid
GROUP BY DistinctUUsersID.user_guid
ORDER BY numrows DESC;
```
**문제를 해결하기 위해 가장 문제를 많이 일으키는 user_guid 'ce7b75bc-7144-11e5-ba71-058fbc01cf0b'과 관련된 행만 살펴 보도록하겠습니다. 위의 쿼리를 다시 쓰되, users테이블에서 user_guid='ce7b75bc-7144-11e5-ba71-058fbc01cf0b'인 *고유의* 사용자 (user)만 left join하세요. 처음 두개의 열은 같은 user_guid가 있고, numrows열의 한 행은 1819가 출력되야 합니다. **

In [28]:
%sql SELECT distinctuserID.user_guid AS uUserID, d.user_guid AS dUserID, count(*) AS numrows\
FROM (SELECT DISTINCT u.user_guid FROM users u WHERE u.user_guid='ce7b75bc-7144-11e5-ba71-058fbc01cf0b') AS distinctuserID LEFT JOIN dogs d\
ON distinctuserID.user_guid=d.user_guid\
GROUP BY distinctuserID.user_guid\
ORDER BY numrows DESC;

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[('ce7b75bc-7144-11e5-ba71-058fbc01cf0b', 'ce7b75bc-7144-11e5-ba71-058fbc01cf0b', 1819)]

In [10]:
%sql SELECT distinctuserID.user_guid AS uUserID, d.user_guid AS dUserID, count(*) AS numrows\
FROM (SELECT DISTINCT u.user_guid FROM users u WHERE u.user_guid='ce7b75bc-7144-11e5-ba71-058fbc01cf0b') AS distinctuserID RIGHT JOIN dogs d\
ON distinctuserID.user_guid=d.user_guid\
GROUP BY distinctuserID.user_guid\
ORDER BY numrows DESC;

 * mysql://studentuser:***@localhost/dognitiondb
2 rows affected.


[(None, None, 33231),
 ('ce7b75bc-7144-11e5-ba71-058fbc01cf0b', 'ce7b75bc-7144-11e5-ba71-058fbc01cf0b', 1819)]

**질문 10: 이제 Innery query를 테스트 해봅시다. dogs 테이블에 임시 이름을 정하고, dogs테이블에서 고유의 user_guids를 고르는 쿼리를 쓰세요. (다음 문제에서 이 쿼리를 inner subquery로 사용할 것이기 때문에 dogs테이블의 user_guid를 user테이블의 user_guid랑 구분할 수 있는 임시 이름이 필요합니다).**  

In [34]:
%sql SELECT COUNT(DISTINCT d.user_guid) FROM dogs d;

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(30967,)]

**질문 11: 질문 10에서 쓴 쿼리를 질문 9의 join의 오른 쪽에 넣으세요. 쿼리 결과는 두 열의 user_guids가 같아야 하고 numrows에는 1이 있어야 합니다** 

In [36]:
%sql SELECT distinctuserID.user_guid AS uUserID, d.user_guid AS dUserID, count(*) AS numrows\
FROM (SELECT DISTINCT u.user_guid FROM users u WHERE u.user_guid='ce7b75bc-7144-11e5-ba71-058fbc01cf0b') AS distinctuserID LEFT JOIN\
(SELECT DISTINCT d.user_guid FROM dogs d) AS d\
ON distinctuserID.user_guid=d.user_guid\
GROUP BY distinctuserID.user_guid\
ORDER BY numrows DESC;

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[('ce7b75bc-7144-11e5-ba71-058fbc01cf0b', 'ce7b75bc-7144-11e5-ba71-058fbc01cf0b', 1)]

**질문 12: 질문 10의 쿼리를 수정해서 user테이블에 있는 user들이 키우는 모든 개들을 출력하세요. 그리고 그 개에 해당하는 품종 관련 정보를 모두 출력하세요. 이 쿼리를 실행하기 전에 Join에 왼쪽에 있는 Subquery의 결과를 100행으로 제한 (LIMIT)하는 것을 잊지 마세요. 이것은 밖에 있는 쿼리의 결과를 100행으로 제한하는 것과 다릅니다. 이렇게 Subquery가 있는 쿼리에서는 Subquery가 제일 먼저 실행되기 때문입니다.**


In [14]:
%sql SELECT u.user_guid, d.dog_guid, d.breed_group FROM (SELECT DISTINCT(user_guid) FROM users AS u LIMIT 100) AS u LEFT JOIN dogs d ON u.user_guid=d.user_guid;

 * mysql://studentuser:***@localhost/dognitiondb
165 rows affected.


[('ce134e42-7144-11e5-ba71-058fbc01cf0b', 'fd27b272-7144-11e5-ba71-058fbc01cf0b', 'Sporting'),
 ('ce134e42-7144-11e5-ba71-058fbc01cf0b', 'fd417cac-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce1353d8-7144-11e5-ba71-058fbc01cf0b', 'fd27b5ba-7144-11e5-ba71-058fbc01cf0b', 'Herding'),
 ('ce1353d8-7144-11e5-ba71-058fbc01cf0b', 'fd3fb0f2-7144-11e5-ba71-058fbc01cf0b', 'Herding'),
 ('ce135ab8-7144-11e5-ba71-058fbc01cf0b', 'fd27b6b4-7144-11e5-ba71-058fbc01cf0b', 'Sporting'),
 ('ce13507c-7144-11e5-ba71-058fbc01cf0b', 'fd27b79a-7144-11e5-ba71-058fbc01cf0b', 'Sporting'),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'fd27b86c-7144-11e5-ba71-058fbc01cf0b', 'Toy'),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'fd27ba1a-7144-11e5-ba71-058fbc01cf0b', 'Toy'),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'fd27e9a4-7144-11e5-ba71-058fbc01cf0b', 'Toy'),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'fd27ed46-7144-11e5-ba71-058fbc01cf0b', 'Toy'),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'fd3cf718-7144-11e5-ba71-058fbc01cf0b', 'Toy'),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'fd3d587a-7144-11e5-ba71-058fbc01cf0b', 'Toy'),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'fd3fbfe8-7144-11e5-ba71-058fbc01cf0b', 'Toy'),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'fd41c400-7144-11e5-ba71-058fbc01cf0b', 'Toy'),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'fd42e33a-7144-11e5-ba71-058fbc01cf0b', 'Toy'),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'fd3cffe2-7144-11e5-ba71-058fbc01cf0b', 'Toy'),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'fd42e196-7144-11e5-ba71-058fbc01cf0b', 'Toy'),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'fd453b6c-7144-11e5-ba71-058fbc01cf0b', 'Toy'),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'fd43c0c0-7144-11e5-ba71-058fbc01cf0b', 'Toy'),
 ('ce13615c-7144-11e5-ba71-058fbc01cf0b', 'fd27b948-7144-11e5-ba71-058fbc01cf0b', 'Working'),
 ('ce135f2c-7144-11e5-ba71-058fbc01cf0b', 'fd27bbbe-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce136a1c-7144-11e5-ba71-058fbc01cf0b', 'fd27c1c2-7144-11e5-ba71-058fbc01cf0b', 'Sporting'),
 ('ce136a1c-7144-11e5-ba71-058fbc01cf0b', 'fd27c0fa-7144-11e5-ba71-058fbc01cf0b', 'Sporting'),
 ('ce136ac6-7144-11e5-ba71-058fbc01cf0b', 'fd27c5be-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce136c24-7144-11e5-ba71-058fbc01cf0b', 'fd27c74e-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce136c24-7144-11e5-ba71-058fbc01cf0b', 'fd27c64a-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce136e36-7144-11e5-ba71-058fbc01cf0b', 'fd27c7d0-7144-11e5-ba71-058fbc01cf0b', 'Sporting'),
 ('ce136ee0-7144-11e5-ba71-058fbc01cf0b', 'fd27c852-7144-11e5-ba71-058fbc01cf0b', 'Toy'),
 ('ce136f94-7144-11e5-ba71-058fbc01cf0b', 'fd27c8d4-7144-11e5-ba71-058fbc01cf0b', 'Working'),
 ('ce136f94-7144-11e5-ba71-058fbc01cf0b', 'fd27cd98-7144-11e5-ba71-058fbc01cf0b', 'Hound'),
 ('ce136f94-7144-11e5-ba71-058fbc01cf0b', 'fd27ce1a-7144-11e5-ba71-058fbc01cf0b', 'Hound'),
 ('ce136f94-7144-11e5-ba71-058fbc01cf0b', 'fd3d249a-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce134be0-7144-11e5-ba71-058fbc01cf0b', 'fd27c956-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce134be0-7144-11e5-ba71-058fbc01cf0b', 'fd3cf7b8-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce134be0-7144-11e5-ba71-058fbc01cf0b', 'fd3cf84e-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce1371a6-7144-11e5-ba71-058fbc01cf0b', 'fd27cb72-7144-11e5-ba71-058fbc01cf0b', 'Hound'),
 ('ce1373ae-7144-11e5-ba71-058fbc01cf0b', 'fd27cea6-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce13750c-7144-11e5-ba71-058fbc01cf0b', 'fd27cf28-7144-11e5-ba71-058fbc01cf0b', 'Sporting'),
 ('ce1375b6-7144-11e5-ba71-058fbc01cf0b', 'fd27cfaa-7144-11e5-ba71-058fbc01cf0b', 'Herding'),
 ('ce1377b4-7144-11e5-ba71-058fbc01cf0b', 'fd27d02c-7144-11e5-ba71-058fbc01cf0b', 'Herding'),
 ('ce137700-7144-11e5-ba71-058fbc01cf0b', 'fd27d0b8-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce137700-7144-11e5-ba71-058fbc01cf0b', 'fd470d8e-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce137868-7144-11e5-ba71-058fbc01cf0b', 'fd27d144-7144-11e5-ba71-058fbc01cf0b', 'Non-Sporting'),
 ('ce137868-7144-11e5-ba71-058fbc01cf0b', 'fd27d1c6-7144-11e5-ba71-0

**질문 13: 이제 왜 dogs와 users테이블에 중복되는 행이 있는지 아시겠나요? 다행이 회사에서 쓰는 데이터에서는 중복되는 데이터가 생성되지 않게 잘 관리가 되어 있는 편입니다. 그래서 확신하려면 위의 쿼리를 다시 한번 살펴봅시다:**

```sql
SELECT DistinctUUsersID.user_guid AS uUserID, d.user_guid AS dUserID, count(*) AS numrows
FROM (SELECT DISTINCT u.user_guid FROM users u) AS DistinctUUsersID 
LEFT JOIN dogs d
  ON DistinctUUsersID.user_guid=d.user_guid
GROUP BY DistinctUUsersID.user_guid
ORDER BY numrows DESC 
```

**위의 쿼리에 dog breed 와 dog weight 열을 출력하세요. 그리고 HAVING절을 사용해서 left join의 결과가 10행 이상이 있는 UserID로 제한해주세요. (답으로 5개의 행이 출력되어야 합니다).**

#### ** HAVING: group by 적용 후 만들어진 테이블 기준으로 조건문 적용

In [19]:
%sql SELECT DistinctUUsersID.user_guid AS uUserID, d.user_guid AS dUserID,d.breed_group, d.weight, count(*) AS numrows\
FROM (SELECT DISTINCT u.user_guid FROM users u) AS DistinctUUsersID LEFT JOIN dogs d\
ON DistinctUUsersID.user_guid=d.user_guid\
GROUP BY DistinctUUsersID.user_guid\
HAVING numrows >= 10\
ORDER BY numrows DESC; 

 * mysql://studentuser:***@localhost/dognitiondb
5 rows affected.


[('ce7b75bc-7144-11e5-ba71-058fbc01cf0b', 'ce7b75bc-7144-11e5-ba71-058fbc01cf0b', 'Toy', 190, 1819),
 ('ce225842-7144-11e5-ba71-058fbc01cf0b', 'ce225842-7144-11e5-ba71-058fbc01cf0b', 'Toy', 190, 26),
 ('ce2258a6-7144-11e5-ba71-058fbc01cf0b', 'ce2258a6-7144-11e5-ba71-058fbc01cf0b', 'Toy', 190, 20),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'ce135e14-7144-11e5-ba71-058fbc01cf0b', 'Toy', 190, 13),
 ('ce29675e-7144-11e5-ba71-058fbc01cf0b', 'ce29675e-7144-11e5-ba71-058fbc01cf0b', None, 60, 11)]

문제의 원인이 되는 사용자 ID는 대부분 체중이 190 파운드인 Shih Tzus인것을 알 수 있음. 이전 연습문제에서 배웠던 것처럼 Dognition 테스트는 이 가짜 사용자를 사용해서 testing을 했기 때문이며 이 사용자들의 데이터는 가짜 데이터임. 회사에서는 쿼리를 실행할때는 테스트 어카운트를 제외하고 분석을 하면 됨. 이렇게 원시 자료에는 다양한 데이터가 있기 때문에 분석전에 데이터세트를 자세히 살펴보는 것이 중요함. 